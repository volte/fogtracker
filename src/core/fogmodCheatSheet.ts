import { LineReader } from '@/utils/lineReader';

export interface Transitions {
  otherArea: string;
  otherEntrance: string;
  thisEntrance?: string;
  originalLine: string;
}

export interface Area {
  areaName: string;
  connections: Transitions[];
}

export interface Connection {
  fromArea: string;
  fromEntrance: string;
  toArea: string;
  toEntrance: string;
  originalLine: string;
}

export interface FogModCheatSheet {
  connections: Connection[];
}

export function parseFogModCheatSheet(cheatSheet: string): FogModCheatSheet {
  const reader = new LineReader(cheatSheet, {
    filter: line => line.trim().length > 0,
  });

  const parseError = (message: string): never => {
    throw new Error(`FogMod cheat sheet error (line ${reader.lineNumber}): ${message}`);
  };

  if (!reader.readLine().startsWith('Seed:')) {
    throw new Error('Invalid FogMod cheat sheet; expected Seed');
  }
  reader.skipLines(2);

  const areas: Area[] = [];
  let line: string | undefined;
  let currentArea: Area | undefined = undefined;
  let finished: boolean = false;

  while ((line = reader.readLine())) {
    if (line.startsWith('Finished')) {
      finished = true;
      break;
    }
    if (line.startsWith('  ')) {
      if (!currentArea) {
        return parseError('transition outside area');
      }
      const matches = line.match(/^ {2}From (\S+) \((.*?)\) to (\S+)(?: \((.*?)\))?/);
      if (!matches) {
        return parseError('invalid transition format');
      }
      currentArea.connections.push({
        otherArea: matches[1],
        thisEntrance: matches[4],
        otherEntrance: matches[2],
        originalLine: line.trim(),
      });
    } else {
      const matches = line.match(/^(\S+)/);
      if (!matches) {
        return parseError('invalid transition format');
      }
      currentArea = { areaName: matches[1], connections: [] };
      areas.push(currentArea);
    }
  }
  if (!finished) {
    return parseError('expected Finished');
  }

  const connections: Connection[] = [];
  for (const area of areas) {
    for (const connection of area.connections) {
      // Don't bother with non-entrance connections, they aren't randomized
      if (!connection.thisEntrance) {
        continue;
      }
      connections.push({
        toArea: area.areaName,
        toEntrance: connection.thisEntrance,
        fromArea: connection.otherArea,
        fromEntrance: connection.otherEntrance,
        originalLine: connection.originalLine,
      });
    }
  }

  return { connections };
}
