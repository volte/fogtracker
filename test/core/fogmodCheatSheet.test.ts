import fs from 'fs';
import { parseFogModCheatSheet } from '@/core/fogmodCheatSheet';

describe('fogmodCheatSheet', () => {
  describe('parseFogModCheatSheet', () => {
    it('parses the cheat sheet', () => {
      const sampleCheatSheetBuffer = fs.readFileSync('test/data/sampleDS1FogModCheatSheet.txt').toString();
      const parsedCheatSheet = parseFogModCheatSheet(sampleCheatSheetBuffer);
      expect(parsedCheatSheet).toMatchSnapshot();
    });
  });
});
