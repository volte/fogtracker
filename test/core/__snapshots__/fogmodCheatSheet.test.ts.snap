// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fogmodCheatSheet parseFogModCheatSheet parses the cheat sheet 1`] = `
{
  "connections": [
    {
      "fromArea": "dukes_caves",
      "fromEntrance": "To Seath in Crystal Caves",
      "originalLine": "From dukes_caves (To Seath in Crystal Caves) to demonruins (In Demon Ruins next to the first Capra Demon)",
      "toArea": "demonruins",
      "toEntrance": "In Demon Ruins next to the first Capra Demon",
    },
    {
      "fromArea": "demonruins",
      "fromEntrance": "Between Quelaag's Domain and the Demon Ruins bonfire",
      "originalLine": "From demonruins (Between Quelaag's Domain and the Demon Ruins bonfire) to parish_gargoyles (To Gargoyles)",
      "toArea": "parish_gargoyles",
      "toEntrance": "To Gargoyles",
    },
    {
      "fromArea": "parish_gargoyles",
      "fromEntrance": "After Gargoyles",
      "originalLine": "From parish_gargoyles (After Gargoyles) to blighttown_quelaag (To Quelaag)",
      "toArea": "blighttown_quelaag",
      "toEntrance": "To Quelaag",
    },
    {
      "fromArea": "blighttown_quelaag",
      "fromEntrance": "After Quelaag",
      "originalLine": "From blighttown_quelaag (After Quelaag) to anorlondo_postrafters (Between the Painting Guardian rafters and the elevator contraption)",
      "toArea": "anorlondo_postrafters",
      "toEntrance": "Between the Painting Guardian rafters and the elevator contraption",
    },
    {
      "fromArea": "anorlondo_painting",
      "fromEntrance": "Using Peculiar Doll to access Painted World",
      "originalLine": "From anorlondo_painting (Using Peculiar Doll to access Painted World) to anorlondo_painting (Leaving Painted World back to Anor Londo)",
      "toArea": "anorlondo_painting",
      "toEntrance": "Leaving Painted World back to Anor Londo",
    },
    {
      "fromArea": "anorlondo_prerafters",
      "fromEntrance": "At the doorway connecting the first Anor Londo elevator and the first gargoyle",
      "originalLine": "From anorlondo_prerafters (At the doorway connecting the first Anor Londo elevator and the first gargoyle) to demonruins_firesage (After Demon Firesage)",
      "toArea": "demonruins_firesage",
      "toEntrance": "After Demon Firesage",
    },
    {
      "fromArea": "anorlondo_main",
      "fromEntrance": "Between the first two Silver Knights and the main building",
      "originalLine": "From anorlondo_main (Between the first two Silver Knights and the main building) to darkroot_sif (To Sif)",
      "toArea": "darkroot_sif",
      "toEntrance": "To Sif",
    },
    {
      "fromArea": "anorlondo_darkmoon",
      "fromEntrance": "To Gwyndolin",
      "originalLine": "From anorlondo_darkmoon (To Gwyndolin) to firelink_gate (Between female Undead Merchant and Lower Burg)",
      "toArea": "firelink_gate",
      "toEntrance": "Between female Undead Merchant and Lower Burg",
    },
    {
      "fromArea": "anorlondo_rafters",
      "fromEntrance": "Between the Painting Guardian rafters and the elevator contraption",
      "originalLine": "From anorlondo_rafters (Between the Painting Guardian rafters and the elevator contraption) to catacombs_pinwheel (Between TotG and Catacombs)",
      "toArea": "catacombs_pinwheel",
      "toEntrance": "Between TotG and Catacombs",
    },
    {
      "fromArea": "firelink",
      "fromEntrance": "Between Firelink Shrine and Upper Burg",
      "originalLine": "From firelink (Between Firelink Shrine and Upper Burg) to burg_posttaurus (Between area after Taurus Demon and Hellkite bridge)",
      "toArea": "burg_posttaurus",
      "toEntrance": "Between area after Taurus Demon and Hellkite bridge",
    },
    {
      "fromArea": "firelink",
      "fromEntrance": "Between Firelink Shrine and Catacombs",
      "originalLine": "From firelink (Between Firelink Shrine and Catacombs) to paintedworld (Between tower with statue and courtyard with phalanx)",
      "toArea": "paintedworld",
      "toEntrance": "Between tower with statue and courtyard with phalanx",
    },
    {
      "fromArea": "paintedworld_priscilla",
      "fromEntrance": "Leaving Painted World back to Anor Londo",
      "originalLine": "From paintedworld_priscilla (Leaving Painted World back to Anor Londo) to paintedworld (Using Peculiar Doll to access Painted World)",
      "toArea": "paintedworld",
      "toEntrance": "Using Peculiar Doll to access Painted World",
    },
    {
      "fromArea": "firelink",
      "fromEntrance": "Between the Firelink Shrine elevator and New Londo Ruins",
      "originalLine": "From firelink (Between the Firelink Shrine elevator and New Londo Ruins) to darkroot (Between Darkroot and Andre)",
      "toArea": "darkroot",
      "toEntrance": "Between Darkroot and Andre",
    },
    {
      "fromArea": "burg_posttaurus",
      "fromEntrance": "After Taurus Demon",
      "originalLine": "From burg_posttaurus (After Taurus Demon) to demonruins_elevator (Between stairways behind Firesage and elevator to Quelaag's Domain)",
      "toArea": "demonruins_elevator",
      "toEntrance": "Between stairways behind Firesage and elevator to Quelaag's Domain",
    },
    {
      "fromArea": "paintedworld",
      "fromEntrance": "To Priscilla",
      "originalLine": "From paintedworld (To Priscilla) to sens_irongolem (From Anor Londo back to Sen's)",
      "toArea": "sens_irongolem",
      "toEntrance": "From Anor Londo back to Sen's",
    },
    {
      "fromArea": "paintedworld",
      "fromEntrance": "Between tower with statue and courtyard with phalanx",
      "originalLine": "From paintedworld (Between tower with statue and courtyard with phalanx) to dlc_start (To Sanctuary Guardian)",
      "toArea": "dlc_start",
      "toEntrance": "To Sanctuary Guardian",
    },
    {
      "fromArea": "darkroot",
      "fromEntrance": "Between Darkroot and tunnel leading to Valley of the Drakes",
      "originalLine": "From darkroot (Between Darkroot and tunnel leading to Valley of the Drakes) to burg_start (Between Firelink Shrine and Upper Burg)",
      "toArea": "burg_start",
      "toEntrance": "Between Firelink Shrine and Upper Burg",
    },
    {
      "fromArea": "darkroot",
      "fromEntrance": "Between the Darkroot Garden bonfire and the area with the stone knights",
      "originalLine": "From darkroot (Between the Darkroot Garden bonfire and the area with the stone knights) to demonruins_lordvessel (To Demon Firesage)",
      "toArea": "demonruins_lordvessel",
      "toEntrance": "To Demon Firesage",
    },
    {
      "fromArea": "darkroot",
      "fromEntrance": "At one end of the bridge at the top of the waterfall with the hydra below",
      "originalLine": "From darkroot (At one end of the bridge at the top of the waterfall with the hydra below) to parish_church (To Gargoyles)",
      "toArea": "parish_church",
      "toEntrance": "To Gargoyles",
    },
    {
      "fromArea": "burg_start",
      "fromEntrance": "Near the drake encounter in Upper Burg",
      "originalLine": "From burg_start (Near the drake encounter in Upper Burg) to dukes_hallway (At the hallway with the crystal warrior in Duke's Archives)",
      "toArea": "dukes_hallway",
      "toEntrance": "At the hallway with the crystal warrior in Duke's Archives",
    },
    {
      "fromArea": "demonruins_lordvessel",
      "fromEntrance": "Golden fog gate in Demon Ruins",
      "originalLine": "From demonruins_lordvessel (Golden fog gate in Demon Ruins) to asylum_outside (Crow transport from Firelink to Asylum)",
      "toArea": "asylum_outside",
      "toEntrance": "Crow transport from Firelink to Asylum",
    },
    {
      "fromArea": "dukes_hallway",
      "fromEntrance": "To Seath in Archives",
      "originalLine": "From dukes_hallway (To Seath in Archives) to demonruins_lower (In Demon Ruins next to the first Capra Demon)",
      "toArea": "demonruins_lower",
      "toEntrance": "In Demon Ruins next to the first Capra Demon",
    },
    {
      "fromArea": "dlc_forest",
      "fromEntrance": "At the top of the elevator from before Artorias to before Manus",
      "originalLine": "From dlc_forest (At the top of the elevator from before Artorias to before Manus) to demonruins_lower (Golden fog gate in Demon Ruins)",
      "toArea": "demonruins_lower",
      "toEntrance": "Golden fog gate in Demon Ruins",
    },
    {
      "fromArea": "demonruins_firesage",
      "fromEntrance": "To Demon Firesage",
      "originalLine": "From demonruins_firesage (To Demon Firesage) to dlc_township (At the bottom of the elevator from before Artorias to before Manus)",
      "toArea": "dlc_township",
      "toEntrance": "At the bottom of the elevator from before Artorias to before Manus",
    },
    {
      "fromArea": "parish_church",
      "fromEntrance": "Between Undead Parish and Andre",
      "originalLine": "From parish_church (Between Undead Parish and Andre) to asylum_oscar (Between hallway with initial weapons and Oscar's cell)",
      "toArea": "asylum_oscar",
      "toEntrance": "Between hallway with initial weapons and Oscar's cell",
    },
    {
      "fromArea": "parish_church",
      "fromEntrance": "In front of Firelink Shrine dual elevator",
      "originalLine": "From parish_church (In front of Firelink Shrine dual elevator) to darkroot_butterfly (To Moonlight Butterfly)",
      "toArea": "darkroot_butterfly",
      "toEntrance": "To Moonlight Butterfly",
    },
    {
      "fromArea": "parish_church",
      "fromEntrance": "At the bridge crossing over the boar in Undead Parish",
      "originalLine": "From parish_church (At the bridge crossing over the boar in Undead Parish) to burg_lower (To Capra Demon)",
      "toArea": "burg_lower",
      "toEntrance": "To Capra Demon",
    },
    {
      "fromArea": "asylum_prebalcony",
      "fromEntrance": "To Asylum Demon from balcony",
      "originalLine": "From asylum_prebalcony (To Asylum Demon from balcony) to dlc_artorias (To Artorias)",
      "toArea": "dlc_artorias",
      "toEntrance": "To Artorias",
    },
    {
      "fromArea": "dukes_library",
      "fromEntrance": "At the hallway with the crystal warrior in Duke's Archives",
      "originalLine": "From dukes_library (At the hallway with the crystal warrior in Duke's Archives) to dlc_artorias (After Artorias)",
      "toArea": "dlc_artorias",
      "toEntrance": "After Artorias",
    },
    {
      "fromArea": "catacombs",
      "fromEntrance": "Between the bridge Patches flips over and the room with two skeletons",
      "originalLine": "From catacombs (Between the bridge Patches flips over and the room with two skeletons) to catacombs (To Pinwheel)",
      "toArea": "catacombs",
      "toEntrance": "To Pinwheel",
    },
    {
      "fromArea": "dlc_township",
      "fromEntrance": "To Manus",
      "originalLine": "From dlc_township (To Manus) to anorlondo_os (After O&S on left side)",
      "toArea": "anorlondo_os",
      "toEntrance": "After O&S on left side",
    },
    {
      "fromArea": "dlc_township",
      "fromEntrance": "Between Gough's stairway and Oolacile Township",
      "originalLine": "From dlc_township (Between Gough's stairway and Oolacile Township) to dukes_caves (Between Duke's Archives and the field before Crystal Caves)",
      "toArea": "dukes_caves",
      "toEntrance": "Between Duke's Archives and the field before Crystal Caves",
    },
    {
      "fromArea": "burg_upper",
      "fromEntrance": "To Taurus Demon",
      "originalLine": "From burg_upper (To Taurus Demon) to blighttown_upper (In between the main pit and wood planks while climbing up/down Blighttown)",
      "toArea": "blighttown_upper",
      "toEntrance": "In between the main pit and wood planks while climbing up/down Blighttown",
    },
    {
      "fromArea": "burg_upper",
      "fromEntrance": "Lower doorway of room with burg bonfire",
      "originalLine": "From burg_upper (Lower doorway of room with burg bonfire) to depths (Near the area with the Channeler in the Depths)",
      "toArea": "depths",
      "toEntrance": "Near the area with the Channeler in the Depths",
    },
    {
      "fromArea": "burg_upper",
      "fromEntrance": "Near the drake encounter in Upper Burg",
      "originalLine": "From burg_upper (Near the drake encounter in Upper Burg) to anorlondo_cathedral (Between the Anor Londo main cathedral and the Giant Blacksmith)",
      "toArea": "anorlondo_cathedral",
      "toEntrance": "Between the Anor Londo main cathedral and the Giant Blacksmith",
    },
    {
      "fromArea": "parish_hellkite",
      "fromEntrance": "Between area after Taurus Demon and Hellkite bridge",
      "originalLine": "From parish_hellkite (Between area after Taurus Demon and Hellkite bridge) to totg_lordvessel (To Nito)",
      "toArea": "totg_lordvessel",
      "toEntrance": "To Nito",
    },
    {
      "fromArea": "parish_hellkite",
      "fromEntrance": "At the bridge crossing over the boar in Undead Parish",
      "originalLine": "From parish_hellkite (At the bridge crossing over the boar in Undead Parish) to dukes_library (At the first elevator in Duke's Archives)",
      "toArea": "dukes_library",
      "toEntrance": "At the first elevator in Duke's Archives",
    },
    {
      "fromArea": "parish_hellkite",
      "fromEntrance": "Upper doorway of room with burg bonfire",
      "originalLine": "From parish_hellkite (Upper doorway of room with burg bonfire) to dukes_machines (Between Duke's Archives and the field before Crystal Caves)",
      "toArea": "dukes_machines",
      "toEntrance": "Between Duke's Archives and the field before Crystal Caves",
    },
    {
      "fromArea": "burg_lower",
      "fromEntrance": "Between female Undead Merchant and Lower Burg",
      "originalLine": "From burg_lower (Between female Undead Merchant and Lower Burg) to demonruins_izalith (To Bed of Chaos)",
      "toArea": "demonruins_izalith",
      "toEntrance": "To Bed of Chaos",
    },
    {
      "fromArea": "burg_lower",
      "fromEntrance": "Between very long ladder and Lower Burg",
      "originalLine": "From burg_lower (Between very long ladder and Lower Burg) to totg_lower (Golden fog gate in Tomb of the Giants)",
      "toArea": "totg_lower",
      "toEntrance": "Golden fog gate in Tomb of the Giants",
    },
    {
      "fromArea": "asylum",
      "fromEntrance": "Between hallway with initial weapons and Oscar's cell",
      "originalLine": "From asylum (Between hallway with initial weapons and Oscar's cell) to dlc_kalameet (To Kalameet)",
      "toArea": "dlc_kalameet",
      "toEntrance": "To Kalameet",
    },
    {
      "fromArea": "catacombs",
      "fromEntrance": "Getting into coffin to Nito",
      "originalLine": "From catacombs (Getting into coffin to Nito) to dlc_sanctuary (After Sanctuary Guardian)",
      "toArea": "dlc_sanctuary",
      "toEntrance": "After Sanctuary Guardian",
    },
    {
      "fromArea": "catacombs",
      "fromEntrance": "Between the bridge Patches flips over and the room with two skeletons",
      "originalLine": "From catacombs (Between the bridge Patches flips over and the room with two skeletons) to anorlondo_gwyndolin (To Gwyndolin)",
      "toArea": "anorlondo_gwyndolin",
      "toEntrance": "To Gwyndolin",
    },
    {
      "fromArea": "sens_irongolem",
      "fromEntrance": "To Iron Golem",
      "originalLine": "From sens_irongolem (To Iron Golem) to demonruins_stairs (Between stairways behind Firesage and elevator to Quelaag's Domain)",
      "toArea": "demonruins_stairs",
      "toEntrance": "Between stairways behind Firesage and elevator to Quelaag's Domain",
    },
    {
      "fromArea": "depths",
      "fromEntrance": "To Gaping Dragon",
      "originalLine": "From depths (To Gaping Dragon) to sens_upper (Between final swinging axes aand Sen's Fortress roof)",
      "toArea": "sens_upper",
      "toEntrance": "Between final swinging axes aand Sen's Fortress roof",
    },
    {
      "fromArea": "darkroot_butterfly",
      "fromEntrance": "After Moonlight Butterfly",
      "originalLine": "From darkroot_butterfly (After Moonlight Butterfly) to sens_upper (To Iron Golem)",
      "toArea": "sens_upper",
      "toEntrance": "To Iron Golem",
    },
    {
      "fromArea": "blighttown_bell",
      "fromEntrance": "After Quelaag",
      "originalLine": "From blighttown_bell (After Quelaag) to catacombs_prepinwheel (To Pinwheel)",
      "toArea": "catacombs_prepinwheel",
      "toEntrance": "To Pinwheel",
    },
    {
      "fromArea": "blighttown_bell",
      "fromEntrance": "Between Quelaag's Domain and the Demon Ruins bonfire",
      "originalLine": "From blighttown_bell (Between Quelaag's Domain and the Demon Ruins bonfire) to dlc_hallway (Between Gough's stairway and Oolacile Township)",
      "toArea": "dlc_hallway",
      "toEntrance": "Between Gough's stairway and Oolacile Township",
    },
    {
      "fromArea": "depths",
      "fromEntrance": "Near the area with the Channeler in the Depths",
      "originalLine": "From depths (Near the area with the Channeler in the Depths) to newlondo (Between the Firelink Shrine elevator and New Londo Ruins)",
      "toArea": "newlondo",
      "toEntrance": "Between the Firelink Shrine elevator and New Londo Ruins",
    },
    {
      "fromArea": "anorlondo_cathedral",
      "fromEntrance": "To O&S",
      "originalLine": "From anorlondo_cathedral (To O&S) to sens_mid (Near the outside boulder stairway and Siegmeyer in Sen's Fortress)",
      "toArea": "sens_mid",
      "toEntrance": "Near the outside boulder stairway and Siegmeyer in Sen's Fortress",
    },
    {
      "fromArea": "anorlondo_cathedral",
      "fromEntrance": "From the Anor Londo main cathedral out of a broken window in front",
      "originalLine": "From anorlondo_cathedral (From the Anor Londo main cathedral out of a broken window in front) to burg_above_bonfire (Upper doorway of room with burg bonfire)",
      "toArea": "burg_above_bonfire",
      "toEntrance": "Upper doorway of room with burg bonfire",
    },
    {
      "fromArea": "anorlondo_cathedral",
      "fromEntrance": "Between the Anor Londo main cathedral and the area with bedrooms and paintings",
      "originalLine": "From anorlondo_cathedral (Between the Anor Londo main cathedral and the area with bedrooms and paintings) to darkroot_stone (Between the Darkroot Garden bonfire and the area with the stone knights)",
      "toArea": "darkroot_stone",
      "toEntrance": "Between the Darkroot Garden bonfire and the area with the stone knights",
    },
    {
      "fromArea": "totg_lordvessel",
      "fromEntrance": "Golden fog gate in Tomb of the Giants",
      "originalLine": "From totg_lordvessel (Golden fog gate in Tomb of the Giants) to darkroot_forest (At one end of the bridge at the top of the waterfall with the hydra below)",
      "toArea": "darkroot_forest",
      "toEntrance": "At one end of the bridge at the top of the waterfall with the hydra below",
    },
    {
      "fromArea": "totg_lower",
      "fromEntrance": "Before the Halberd Black Knight in TotG",
      "originalLine": "From totg_lower (Before the Halberd Black Knight in TotG) to anorlondo_start (From Sen's to Anor Londo)",
      "toArea": "anorlondo_start",
      "toEntrance": "From Sen's to Anor Londo",
    },
    {
      "fromArea": "newlondo",
      "fromEntrance": "Near the New Londo Ruins shortcut ladder",
      "originalLine": "From newlondo (Near the New Londo Ruins shortcut ladder) to newlondo_drakes (Between Blighttown and Valley of the Drakes)",
      "toArea": "newlondo_drakes",
      "toEntrance": "Between Blighttown and Valley of the Drakes",
    },
    {
      "fromArea": "burg_bonfire",
      "fromEntrance": "Lower doorway of room with burg bonfire",
      "originalLine": "From burg_bonfire (Lower doorway of room with burg bonfire) to demonruins_centipede (To Centipede Demon)",
      "toArea": "demonruins_centipede",
      "toEntrance": "To Centipede Demon",
    },
    {
      "fromArea": "dlc_sanctuary",
      "fromEntrance": "Between Sanctuary Garden and Royal Forest",
      "originalLine": "From dlc_sanctuary (Between Sanctuary Garden and Royal Forest) to dlc_forest (To Kalameet)",
      "toArea": "dlc_forest",
      "toEntrance": "To Kalameet",
    },
    {
      "fromArea": "dlc_forest",
      "fromEntrance": "Between Sanctuary Garden and Royal Forest",
      "originalLine": "From dlc_forest (Between Sanctuary Garden and Royal Forest) to dlc_forest (To Artorias)",
      "toArea": "dlc_forest",
      "toEntrance": "To Artorias",
    },
    {
      "fromArea": "demonruins_stairs",
      "fromEntrance": "After Demon Firesage",
      "originalLine": "From demonruins_stairs (After Demon Firesage) to anorlondo_blacksmith (Between the Anor Londo main cathedral and the Giant Blacksmith)",
      "toArea": "anorlondo_blacksmith",
      "toEntrance": "Between the Anor Londo main cathedral and the Giant Blacksmith",
    },
    {
      "fromArea": "demonruins_stairs",
      "fromEntrance": "To Centipede Demon",
      "originalLine": "From demonruins_stairs (To Centipede Demon) to paintedworld_priscilla (To Priscilla)",
      "toArea": "paintedworld_priscilla",
      "toEntrance": "To Priscilla",
    },
    {
      "fromArea": "dlc_hallway",
      "fromEntrance": "Between Gough's stairway and Artorias",
      "originalLine": "From dlc_hallway (Between Gough's stairway and Artorias) to dlc_guardian (After Sanctuary Guardian)",
      "toArea": "dlc_guardian",
      "toEntrance": "After Sanctuary Guardian",
    },
    {
      "fromArea": "blighttown",
      "fromEntrance": "To Quelaag",
      "originalLine": "From blighttown (To Quelaag) to dlc_bottom (At the bottom of the elevator from before Artorias to before Manus)",
      "toArea": "dlc_bottom",
      "toEntrance": "At the bottom of the elevator from before Artorias to before Manus",
    },
    {
      "fromArea": "blighttown",
      "fromEntrance": "Between Blighttown and Valley of the Drakes",
      "originalLine": "From blighttown (Between Blighttown and Valley of the Drakes) to greathollow_ashlake (Between Great Hollow and Ash Lake)",
      "toArea": "greathollow_ashlake",
      "toEntrance": "Between Great Hollow and Ash Lake",
    },
    {
      "fromArea": "blighttown",
      "fromEntrance": "Between Blighttown and Great Hollow",
      "originalLine": "From blighttown (Between Blighttown and Great Hollow) to dlc_manus (To Manus)",
      "toArea": "dlc_manus",
      "toEntrance": "To Manus",
    },
    {
      "fromArea": "sens_mid",
      "fromEntrance": "Between final swinging axes aand Sen's Fortress roof",
      "originalLine": "From sens_mid (Between final swinging axes aand Sen's Fortress roof) to parish_bell (After Gargoyles)",
      "toArea": "parish_bell",
      "toEntrance": "After Gargoyles",
    },
    {
      "fromArea": "darkroot_stone",
      "fromEntrance": "To Moonlight Butterfly",
      "originalLine": "From darkroot_stone (To Moonlight Butterfly) to demonruins_ceaseless (To Ceaseless Discharge)",
      "toArea": "demonruins_ceaseless",
      "toEntrance": "To Ceaseless Discharge",
    },
    {
      "fromArea": "darkroot_forest",
      "fromEntrance": "To Sif",
      "originalLine": "From darkroot_forest (To Sif) to greathollow (Between Blighttown and the Great Hollow)",
      "toArea": "greathollow",
      "toEntrance": "Between Blighttown and the Great Hollow",
    },
    {
      "fromArea": "anorlondo_start",
      "fromEntrance": "Golden fog gate leading to Duke's Archives",
      "originalLine": "From anorlondo_start (Golden fog gate leading to Duke's Archives) to asylum_balcony (To Asylum Demon from balcony)",
      "toArea": "asylum_balcony",
      "toEntrance": "To Asylum Demon from balcony",
    },
    {
      "fromArea": "newlondo_drakes",
      "fromEntrance": "Between Darkroot and tunnel leading to Valley of the Drakes",
      "originalLine": "From newlondo_drakes (Between Darkroot and tunnel leading to Valley of the Drakes) to firelink_crow (Warp to Firelink crow's nest)",
      "toArea": "firelink_crow",
      "toEntrance": "Warp to Firelink crow's nest",
    },
    {
      "fromArea": "demonruins_izalithprelava",
      "fromEntrance": "After Centipede Demon",
      "originalLine": "From demonruins_izalithprelava (After Centipede Demon) to newlondo_building (Near the New Londo Ruins shortcut ladder)",
      "toArea": "newlondo_building",
      "toEntrance": "Near the New Londo Ruins shortcut ladder",
    },
    {
      "fromArea": "dlc_top",
      "fromEntrance": "At the top of the elevator from before Artorias to before Manus",
      "originalLine": "From dlc_top (At the top of the elevator from before Artorias to before Manus) to totg_nito (To Nito)",
      "toArea": "totg_nito",
      "toEntrance": "To Nito",
    },
    {
      "fromArea": "anorlondo_os",
      "fromEntrance": "To O&S",
      "originalLine": "From anorlondo_os (To O&S) to dukes_seath2 (To Seath in Crystal Caves)",
      "toArea": "dukes_seath2",
      "toEntrance": "To Seath in Crystal Caves",
    },
    {
      "fromArea": "anorlondo_elevator1",
      "fromEntrance": "After O&S on left side",
      "originalLine": "From anorlondo_elevator1 (After O&S on left side) to depths_gaping (To Gaping Dragon)",
      "toArea": "depths_gaping",
      "toEntrance": "To Gaping Dragon",
    },
    {
      "fromArea": "sens",
      "fromEntrance": "Near the outside boulder stairway and Siegmeyer in Sen's Fortress",
      "originalLine": "From sens (Near the outside boulder stairway and Siegmeyer in Sen's Fortress) to anorlondo_building (Between the first two Silver Knights and the main building)",
      "toArea": "anorlondo_building",
      "toEntrance": "Between the first two Silver Knights and the main building",
    },
    {
      "fromArea": "greathollow",
      "fromEntrance": "Between Great Hollow and Ash Lake",
      "originalLine": "From greathollow (Between Great Hollow and Ash Lake) to darkroot_postbutterfly (After Moonlight Butterfly)",
      "toArea": "darkroot_postbutterfly",
      "toEntrance": "After Moonlight Butterfly",
    },
    {
      "fromArea": "demonruins_centipede",
      "fromEntrance": "After Centipede Demon",
      "originalLine": "From demonruins_centipede (After Centipede Demon) to parish_andre (Between Undead Parish and Andre)",
      "toArea": "parish_andre",
      "toEntrance": "Between Undead Parish and Andre",
    },
    {
      "fromArea": "asylum_stray",
      "fromEntrance": "In Stray Demon arena",
      "originalLine": "From asylum_stray (In Stray Demon arena) to burg_taurus (To Taurus Demon)",
      "toArea": "burg_taurus",
      "toEntrance": "To Taurus Demon",
    },
    {
      "fromArea": "anorlondo_building",
      "fromEntrance": "Between the Anor Londo main cathedral and the area with bedrooms and paintings",
      "originalLine": "From anorlondo_building (Between the Anor Londo main cathedral and the area with bedrooms and paintings) to totg_coffin (Getting into coffin to Nito)",
      "toArea": "totg_coffin",
      "toEntrance": "Getting into coffin to Nito",
    },
    {
      "fromArea": "dlc_guardian",
      "fromEntrance": "To Sanctuary Guardian",
      "originalLine": "From dlc_guardian (To Sanctuary Guardian) to totg (Between TotG and Catacombs)",
      "toArea": "totg",
      "toEntrance": "Between TotG and Catacombs",
    },
    {
      "fromArea": "demonruins",
      "fromEntrance": "To Ceaseless Discharge",
      "originalLine": "From demonruins (To Ceaseless Discharge) to newlondo_stairs (To Four Kings)",
      "toArea": "newlondo_stairs",
      "toEntrance": "To Four Kings",
    },
    {
      "fromArea": "blighttown",
      "fromEntrance": "In between the main pit and wood planks while climbing up/down Blighttown",
      "originalLine": "From blighttown (In between the main pit and wood planks while climbing up/down Blighttown) to newlondo_lower (On an entrance to the building with the Very Large Ember in lower ruins)",
      "toArea": "newlondo_lower",
      "toEntrance": "On an entrance to the building with the Very Large Ember in lower ruins",
    },
    {
      "fromArea": "anorlondo_start",
      "fromEntrance": "At the doorway connecting the first Anor Londo elevator and the first gargoyle",
      "originalLine": "From anorlondo_start (At the doorway connecting the first Anor Londo elevator and the first gargoyle) to newlondo_lower (On an entrance to the building with the Very Large Ember in lower ruins)",
      "toArea": "newlondo_lower",
      "toEntrance": "On an entrance to the building with the Very Large Ember in lower ruins",
    },
    {
      "fromArea": "parish_andre",
      "fromEntrance": "Between Darkroot and Andre",
      "originalLine": "From parish_andre (Between Darkroot and Andre) to catacombs_start (Between Firelink Shrine and Catacombs)",
      "toArea": "catacombs_start",
      "toEntrance": "Between Firelink Shrine and Catacombs",
    },
    {
      "fromArea": "parish_andre",
      "fromEntrance": "Between Sen's Fortress gate and Andre",
      "originalLine": "From parish_andre (Between Sen's Fortress gate and Andre) to dukes_start (Golden fog gate leading to Duke's Archives)",
      "toArea": "dukes_start",
      "toEntrance": "Golden fog gate leading to Duke's Archives",
    },
    {
      "fromArea": "newlondo_lower",
      "fromEntrance": "To Four Kings",
      "originalLine": "From newlondo_lower (To Four Kings) to dukes_start (At the first elevator in Duke's Archives)",
      "toArea": "dukes_start",
      "toEntrance": "At the first elevator in Duke's Archives",
    },
    {
      "fromArea": "sens",
      "fromEntrance": "Between Sen's Fortress gate and Andre",
      "originalLine": "From sens (Between Sen's Fortress gate and Andre) to firelink_elevator (In front of Firelink Shrine dual elevator)",
      "toArea": "firelink_elevator",
      "toEntrance": "In front of Firelink Shrine dual elevator",
    },
    {
      "fromArea": "totg",
      "fromEntrance": "Before the Halberd Black Knight in TotG",
      "originalLine": "From totg (Before the Halberd Black Knight in TotG) to dukes_seath1 (To Seath in Archives)",
      "toArea": "dukes_seath1",
      "toEntrance": "To Seath in Archives",
    },
    {
      "fromArea": "burg_ladder",
      "fromEntrance": "Between very long ladder and Lower Burg",
      "originalLine": "From burg_ladder (Between very long ladder and Lower Burg) to burg_capra (To Capra Demon)",
      "toArea": "burg_capra",
      "toEntrance": "To Capra Demon",
    },
    {
      "fromArea": "burg_taurus",
      "fromEntrance": "After Taurus Demon",
      "originalLine": "From burg_taurus (After Taurus Demon) to demonruins_bedofchaos (To Bed of Chaos)",
      "toArea": "demonruins_bedofchaos",
      "toEntrance": "To Bed of Chaos",
    },
  ],
}
`;
